### PR TITLE
 gw deleting state

### PR DESCRIPTION
**Pull request by leonidc**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/mon/NVMeofGwMap.cc [hunk 1]
```diff
@@ -78,10 +81,37 @@ int NVMeofGwMap::cfg_add_gw(
...
     if (itr.first == gw_id) {
-      dout(1) << __func__ << " ERROR create GW: already exists in map "
-	      << gw_id << dendl;
-      return -EEXIST ;
+      if (itr.second.availability  != gw_availability_t::GW_DELETING) {
+        dout(1) << __func__ << " ERROR create GW: already exists in map "
+	          << gw_id << dendl;
+        return -EEXIST ;
+      }
+      else {
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.cc:89]: } else {

Check every instance.

**Comment by athanatos** [comment]
@leonidc Can you explain what the bug being fixed actually is in the commit message?  Doesn't need to be a lot of detail, just enough to understand the diff and what the new state is supposed to represent.

File: src/mon/NVMeofGwMap.h [hunk 2]
```diff
@@ -44,7 +44,8 @@ class NVMeofGwMap
   std::map<NvmeGroupKey, NvmeGwTimers> fsm_timers;
 
   void to_gmap(std::map<NvmeGroupKey, NvmeGwMonClientStates>& Gmap) const;
-
+  void  track_deleting_gws            (const NvmeGroupKey& group_key,
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.h:47]: Extra space after void, remove spaces between track_deleting_gws and (

File: src/mon/NVMeofGwMap.h [hunk 3]
```diff
@@ -95,7 +99,8 @@ class NVMeofGwMap
   void set_failover_gw_for_ANA_group(
     const NvmeGwId &failed_gw_id, const NvmeGroupKey& group_key,
     const NvmeGwId &gw_id, NvmeAnaGrpId groupid);
-
+  int  get_num_namespaces (const NvmeGwId &gw_id,
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.h:102]: Extra space after int, extra space between namespaces and (

**Review by athanatos — CHANGES_REQUESTED** [review]
Functionality looks probably ok after formatting fixes are addressed.  Ping me once PR is updated and I'll kick off a test run.

**Comment by athanatos** [comment]
Update commit message with (brief) description of the bug being fixed.

File: src/mon/NVMeofGwMap.cc [hunk 4]
```diff
@@ -36,10 +36,13 @@ void NVMeofGwMap::to_gmap(
     const NvmeGwMonStates& gw_created_map = created_map_pair.second;
     for (const auto& gw_created_pair: gw_created_map) {
       const auto& gw_id = gw_created_pair.first;
-      const auto& gw_created  = gw_created_pair.second;
+      auto& gw_created  = gw_created_pair.second;
```
[the remark this answers is not in the document]
Comment (leonidc) [reply src/mon/NVMeofGwMap.cc:39]: fixing

File: src/mon/NVMeofGwMon.cc [hunk 5]
```diff
@@ -391,16 +428,17 @@ bool NVMeofGwMon::prepare_command(MonOpRequestRef op)
     }
     else{
       rc = pending_map.cfg_delete_gw(id, group_key);
-      if (rc == -EINVAL) {
+      if (rc == 0) {
+        bool propose = false;
+        // Simulate  immediate Failover of this GW
+        process_gw_down(id, group_key, propose);
+      }
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/mon/NVMeofGwMon.cc:436]: This has not been addressed.

File: src/mon/NVMeofGwMap.cc [hunk 6]
```diff
@@ -37,9 +37,14 @@ void NVMeofGwMap::to_gmap(
     for (const auto& gw_created_pair: gw_created_map) {
       const auto& gw_id = gw_created_pair.first;
       const auto& gw_created  = gw_created_pair.second;
+      gw_availability_t availability = gw_created.availability;
+      // Gateways expect to see UNAVAILABLE, not DELETING
+      // for entries in DELETING state
+      if (gw_created.availability == gw_availability_t::GW_DELETING)
+         availability = gw_availability_t::GW_UNAVAILABLE;
```
Comment (athanatos) [inline src/mon/NVMeofGwMap.cc:44]: Braces around if body.

**Review by athanatos — APPROVED** [review]
One of my comments had not been addressed, but since you've presumably already tested this version we can leave it as it is -- please check over your PRs for typos in the future before asking for review.

**Comment by athanatos** [comment]
jenkins test api

**Comment by athanatos** [comment]
@leonidc You still have not fixed the commit message.

**Comment by leonidc** [comment]
jenkins api test

**Comment by leonidc** [comment]
jenkins api test

**Comment by caroav** [comment]
jenkins test api

**Pull request merged**